extensions/cornerstone/src/tools/ImageOverlayViewerTool.tsx

```diff
@@ -0,0 +1,258 @@
+ * But it is used to display Pixel Overlays. And it will provide toggling capability.
+ *
+ * The documentation for Overlay Plane Module of DICOM can be found in [C.9.2 of
+ * Part-3 of DICOM standard](https://dicom.nema.org/medical/dicom/2018b/output/chtml/part03/sect_C.9.2.html)
+ */
+class ImageOverlayViewerTool extends BaseTool {
+  static toolName = 'ImageOverlayViewer';
+
+  _renderingViewport: any;
+  _cachedStats: { [key: string]: CachedStat } = {};
+
+  constructor(
+    toolProps = {},
+    defaultToolProps = {
+      supportedInteractionTypes: [],
+      configuration: {
+        showOverlays: true,
+        fillColor: [255, 127, 127, 255],
+      },
+    }
+  ) {
+    super(toolProps, defaultToolProps);
+  }
+
+  private _getCachedStat(targetId: string, createIfMissing = true): CachedStat {
+    if (!this._cachedStats[targetId] && createIfMissing)
+      this._cachedStats[targetId] = {
+        color: [127, 127, 127, 255], // color (r,g,b,a), default color: gray
+        overlays: [],
+      };
+    return this._cachedStats[targetId];
+  }
+
+  renderAnnotation = (enabledElement, svgDrawingHelper) => {
+    // overlays are toggled off by configuration
+    if (!this.configuration.showOverlays) return false;
+
+    const { viewport } = enabledElement;
+    this._renderingViewport = viewport;
+
+    const imageId = this._getReferencedImageId(viewport);
+    if (!imageId) return;
+
+    const targetId = this.getTargetId(viewport);
+    const cachedStat = this._getCachedStat(targetId);
+
+    let overlays;
+    ({ overlays } = metaData.get('overlayPlaneModule', imageId));
+    overlays = overlays.filter(overlay => overlay.type === 'G');
+
+    // no overlays
+    if (!overlays || overlays.length <= 0) return;
+
+    overlays.forEach(async (overlay, idx) => {
+      try {
+        if (!cachedStat.overlays[idx]) {
+          // first time load? let's load data
+          if (overlay.pixelData) {
+            let pixelData = null;
+            if (overlay.pixelData.Value) {
+              pixelData = overlay.pixelData.Value;
+            } else if (overlay.pixelData.retrieveBulkData) {
+              pixelData = await overlay.pixelData.retrieveBulkData();
+            }
+
+            if (pixelData) {
+              cachedStat.overlays[idx] = {
+                ...overlay,
+                pixelDataRaw: pixelData, // this will be an ArrayBuffer object
+                _id: guid(),
+              };
+            }
+          }
+        }
+
+        if (cachedStat.overlays[idx]) {
+          this._renderOverlay(
+            enabledElement,
+            svgDrawingHelper,
+            cachedStat.overlays[idx]
+          );
+        }
+      } catch (e) {
+        console.error('Failed to render overlay', e);
+      }
+    });
+
+    return true;
+  };
+
+  private _renderOverlay(enabledElement, svgDrawingHelper, overlayData) {
+    if (
+      !overlayData.color ||
+      this.configuration.fillColor != overlayData.color
+    ) {
```

Comment:
Please add the actual generation of the dataUrl into a separate function and store the dataUrl as the primary output of the overall asynchronous function which loads the overlay data.  That way, you cache just the dataUrl, and turning it on/off is just whether you include the SVG element or not.  Also, that generation should probably occur as part of general image loading, which separates out the timing from the creation of the tool object.